src/lib/gossip_net/gossip_net.ml

```diff
@@ -180,30 +252,70 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
        but it only works on Linux. I think we don't need to worry about it,
        there are a lot of ephemeral ports and we don't make very many
        simultaneous connections.
-    *)
-    Async.Socket.bind_inet socket @@ `Inet (t.addrs_and_ports.bind_ip, 0)
+      *)
+      Async.Socket.bind_inet unbound_socket
+      @@ `Inet (t.addrs_and_ports.bind_ip, 0)
+    in
+    let peer_addr = `Inet (peer.Peer.host, peer.communication_port) in
+    try_with (fun () ->
+        let%map connected_socket = Async.Socket.connect socket peer_addr in
+        let fd = Socket.fd connected_socket in
+        (Reader.create fd, Writer.create fd) )
+
+  let close_connection reader writer =
+    Writer.close writer ~force_close:(Clock.after (sec 30.))
+    >>= fun () -> Reader.close reader
+
+  let collect_errors writer f =
+    let monitor = Writer.monitor writer in
+    ignore (Monitor.detach_and_get_error_stream monitor) ;
+    choose
+      [ choice (Monitor.get_next_error monitor) (fun e -> Error e)
+      ; choice (try_with ~name:"Tcp.collect_errors" f) Fn.id ]
+
+  (* see if we can connect to a disconnected peer, every so often *)
+  let retry_disconnected_peer t =
+    let rec loop () =
+      let%bind () = Async.after (Time.Span.of_sec 30.0) in
```

Comment:
this should have exponential backoff with some random jitter

Reply:
but maybe we should just prioritize getting this in, and we can address that with an issue? Does anyone else think this will cause any problems without something like exponential backoff?

Reply:
It's pretty infrequent, every 30 seconds, and only happens when there are no peers available, so you're not pre-empting any other action. Also, the action taken is just a connection attempt, so you're not consuming much bandwidth.